src/traces/image/attributes.js

```diff
@@ -0,0 +1,84 @@
+/**
+* Copyright 2012-2019, Plotly, Inc.
+* All rights reserved.
+*
+* This source code is licensed under the MIT license found in the
+* LICENSE file in the root directory of this source tree.
+*/
+
+'use strict';
+
+var plotAttrs = require('../../plots/attributes');
+var hovertemplateAttrs = require('../../plots/template_attributes').hovertemplateAttrs;
+var extendFlat = require('../../lib/extend').extendFlat;
+
+module.exports = extendFlat({
+    z: {
+        valType: 'data_array',
+        role: 'info',
+        editType: 'calc',
+        description: [
+            'A 2-dimensional array in which each element is an array of 3 or 4 numbers representing a color.',
+        ].join(' ')
+    },
+    colormodel: {
+        valType: 'enumerated',
+        values: ['rgb', 'rgba', 'hsl', 'hsla'],
+        dflt: 'rgb',
+        role: 'info',
+        editType: 'plot',
+        description: 'Color model used to map the numerical color components described in `z` into colors.'
+    },
+    zmin: {
+        valType: 'data_array',
+        role: 'info',
+        editType: 'plot',
+        description: [
+            'Array defining the lower bound for each color component.',
+            'For example, for the `rgba` colormodel, the default value is [0, 0, 0, 0].'
+        ].join(' ')
+    },
+    zmax: {
+        valType: 'data_array',
+        role: 'info',
+        editType: 'plot',
+        description: [
+            'Array defining the higher bound for each color component.',
+            'For example, for the `rgba` colormodel, the default value is [255, 255, 255, 1].'
+        ].join(' ')
+    },
+    x0: {
+        valType: 'number',
+        dflt: 0,
+        role: 'info',
+        editType: 'calc',
+        description: 'Set the image\'s x position.'
+    },
+    y0: {
+        valType: 'number',
+        dflt: 0,
+        role: 'info',
+        editType: 'calc',
+        description: 'Set the image\'s y position.'
+    },
+    dx: {
+        valType: 'number',
+        dflt: 1,
+        role: 'info',
+        editType: 'calc',
+        description: 'Set the pixel\'s horizontal size.'
+    },
+    dy: {
+        valType: 'number',
+        dflt: 1,
+        role: 'info',
+        editType: 'calc',
+        description: 'Set the pixel\'s vertical size'
+    },
+    hoverinfo: extendFlat({}, plotAttrs.hoverinfo, {
+        flags: ['x', 'y', 'z', 'color', 'name']
+    }),
+    hovertemplate: hovertemplateAttrs({}, {
+        keys: ['z', 'c', 'colormodel']
```

Comment:
What's `c` here? Is it the same value as for the `'color'` hoverinfo flag?

Reply:
... moreover, do we really need `colormodel` in here? It's a per-trace attribute, so it should be available using:

```js
hovertemplate: 'using color model: %{fullData.colormodel}'
```

Reply:
> What's `c` here? Is it the same value as for the `'color'` hoverinfo flag?

`c` is the color array. When using the `'color'` hoverinfo flag, we display the `c` array WITH the colormodel in front of it ie. `%{colormodel}: [%{c[0]}, %{c[1]}, %{c[2]}]`. Should I rename `c` to `color` for `hovertemplate`?

Reply:
So currently,

```js
Plotly.newPlot(gd, [{
  type: 'image',
  z: [[ [255,0,0] ]],
  hoverinfo: 'color'
}])

// => shows `RBG: [255, 0, 0]` on hover
```

But

```js
Plotly.newPlot(gd, [{
  type: 'image',
  z: [[ [255,0,0] ]],
  hovertemplate: '%{color}'

// => shows `%{color}` 
}])
```

... which I'm not a fan of. I think we should try to make `hoverinfo` flags and `hovertemplate` token match. So yeah, I think renaming `c` to `color` is the way to go.

---

Moreover, I really like that `%{c[0]}`, `%{c[1]}` and friends (soon to be renamed `color[0]`, `color[1]`, ...) work, but it would nice to make

```js
Plotly.newPlot(gd, [{
  type: 'image',
  z: [[ [255,0,0] ]],
  hovertemplate: '%{c[3]}'

// => currently shows `%{c[3]}` 
}])
```

show `1` instead, even when the alpha channel isn't set.

Reply:
I renamed `c` to `color` in a98ac44ae81685920979202770200a1e2a05fcba.

Let me think about what to do for `%{color[3]}`. I'm tempted to leave it as is because, in the context of `rgb` or `hsl`, the 4th channel is undefined and shouldn't exist.

Reply:
> Let me think about what to do for `%{color[3]}`. I'm tempted to leave it as is because, in the context of `rgb` or `hsl`, the 4th channel is undefined and shouldn't exist.

That's fair. I think we should show _something_ for `%{color[3]}`, rendering it as just `%{color[3]}` makes me think that the token is invalid, which isn't really true. In my mind, it's the channel that's undefined. So I'd vote for showing any of: `null`,`false`, `''`, `1` or `undefined`. 

But this is a very tiny detail, so feel free to simply add a `TODO` in the code about it and move on to more important things.

Reply:
This one is actually tricky to implement. I would need to set `color[3]` to `1` but then it would show up in hover template `%{color}` which we don't want. The template string routine does not allow overriding the result of accessing an individual array element without ALSO overriding the result when accessing the whole array...

Reply:
I added a `TODO` in the code in commit https://github.com/plotly/plotly.js/commit/e4cac27013813e94a4439809d2de1c53e90c0f95

Reply:
Thanks! Resolving!